### DISCUSSION TITLE
Recorded Midi is _ahead_ of tick

### DISCUSSION THREAD

**Discussion by jhamb** [hydrogen-music/hydrogen#1851] [Hydrogen on Linux]
Hi,

I am facing a strange problem while recording patterns in hydrogen (1.2.0-beta, debian bookwork, kernel 6.x) via midi:
The Midi notes recorded are _ahead_ of the beat. Here you have to believe me, it is not fault of my bad drum timing.
This is what it looks like after recording:
![image](https://github.com/hydrogen-music/hydrogen/assets/301638/84869a1a-eb0d-446b-ac72-cd02336a16eb)

It does not matter whether I use hydrogen as jack master or not. Or whether i use jack transport while recording at all.
![image](https://github.com/hydrogen-music/hydrogen/assets/301638/a57680e8-8ffa-406b-a1b0-c879f674b2a2)

The offset seems to be quite static. 
Anyone an idea why this happens and how to avoid this? I would understand if it had been the other way: notes were lagging due to latency. But it is the other way - notes are ahead as if some component would try to compensate the expected latency lag.

Any hints are welcome!

**Comment by theGreatWhiteShark** [comment]
Hey @jhamb ,

> Anyone an idea why this happens

Hydrogen offers humanization of notes by randomly shifting their onset forward or backward in time. For this to be possible Hydrogen uses two different transport positions. One ahead of time, which is used to pick up and enqueue the notes, and one at the current time, which is used for playback. The difference between the two is depending on sample rate and tempo but is static in case these two aren't altered.

While recording MIDI notes those are gathered "ahead of time" and are placed on their proper position (in the past).

>  how to avoid this?

In that very code of note placement a fix had been introduced since 1.2.0-beta. Not sure whether this already fixes your issue but I would suggest you try using the [latest version](https://github.com/hydrogen-music/hydrogen/releases/download/1.2.2/Hydrogen-1.2.2-x86_64.AppImage) first.

**Reply by jhamb** [reply]
Hi @theGreatWhiteShark ,
thank you very much for your input.

I have tried the AppImage version (1.2.2). The problem still exists. I have the feeling the offset is now slightly shorter but that might be also some kind of placebo.
"Quantize events to grid" makes the problem a bit worse because it "snaps" the midi notes to even earlier grid point in most cases. If it matters: jack runs at 48kHz.

I record midi to hydrogen metronome click and I am pretty sure I am not ahead of the click at all times while recording.
BPM is set to 120. The note is ahead of tick about 2 * 1/32 (please see the screenshot for position of recorded notes in hydrogen pattern). 
This means:
the pattern is 2 secs long and recorded note offset is 2sec / 32 * 2 = 0.125sec. That's way too far from the 5.33msec process time reported by jack. 

Any other ideas how to nail it down are welcome!

**Reply by theGreatWhiteShark** [reply]
Alright. I probably need a couple of days but I try to reproduce the issue and come up with a fix or at least with a debug AppImage that might help us pin it down.

**Reply by jhamb** [reply]
Just tested recording to ardour midi track and route this midi out to hydrogen midi in (just for playing notes in hydrogen, not recording). No latency for me at all. So this is somehow specific to hydrogen. Probably I will need to go with this setup for now. 
I am happy to help to debug this issue if it is interesting for hydrogen devs. Even more happy I would become if I can record midi directly to hydrogen. I find hydrogen pattern editing way more comfortable.

**Reply by theGreatWhiteShark** [reply]
> I am happy to help to debug this issue if it is interesting for hydrogen devs.

Of course it is. Could you check whether [this AppImage](https://ci.appveyor.com/api/buildjobs/tl7u64fujja5huw0/artifacts/Hydrogen-x86_64.AppImage) solves your recording issue?

**Reply by jhamb** [reply]
Just gave it a try - it seems this version solves the issue! Thank you for providing a fix!
